feat: teach DeltaArray slice and scalar_at

danking opened the pull request:
More subtle than I expected.

DeltaArray is a sequence of chunks. All chunks except the last must be "full", i.e. containing 1,024 values. The last chunk may contain as few as one value and is encoded differently from the rest.

In this PR, I introduced an "offset" and a "limit". Together they enable logical/lazy slicing while preserving full chunks for later decompression. The offset is a value, less than 1024, which offsets into the first chunk. The limit is either `None` or less than 1024. `None` represents no limit which allows callers to avoid computing the length of the last chunk [1]. Internally, the limit is converted to a "right-offset": `trailing_garbage` which is sliced away when decompression happens.

I also added support for signed values. I verify they're non-negative in `delta_compress`.

[1] Which is, a bit annoyingly, this:

```
match deltas.len() % 1024 {
    0 => 1024,
    n => n
}
```

danking committed:
feat: teach DeltaArray slice and scalar_at

More subtle than I expected.

DeltaArray is a sequence of chunks. All chunks except the last must be "full", i.e. containing 1,024
values. The last chunk may contain as few as one value and is encoded differently from the rest.

In this PR, I introduced an "offset" and a "limit". Together they enable logical/lazy slicing while
preserving full chunks for later decompression. The offset is a value, less than 1024, which offsets
into the first chunk. The limit is either `None` or less than 1024. `None` represents no limit which
allows callers to avoid computing the length of the last chunk [1]. Internally, the limit is
converted to a "right-offset": `trailing_garbage` which is sliced away when decompression happens.

[1] Which is, a bit annoyingly, this:

```
match deltas.len() % 1024 {
    0 => 1024,
    n => n
}
```

robert3005 commented on encodings/fastlanes/src/delta/compress.rs:
```diff
@@ -3,14 +3,27 @@ use fastlanes::{Delta, Transpose};
 use crate::DeltaArray;
 
 pub fn delta_compress(array: &PrimitiveArray) -> VortexResult<(PrimitiveArray, PrimitiveArray)> {
+    let array = if array.ptype().is_signed_int() {
+        match_each_signed_integer_ptype!(array.ptype(), |$T| {
+            assert!(array.statistics().compute_min::<$T>().map(|x| x >= 0).unwrap_or(false));
+        });
+        &array.reinterpret_cast(array.ptype().to_unsigned())
+    } else {
+        array
+    };
+
```
Generally can handle this in the compressor where you would say you will only delta compress unsigned types and then we FoR compressor should kick in to convert things into unsigned and you will only ever see unsinged integers

robert3005 replied:
The compressor actually already does this.

danking replied:
Done.

danking commented:
If you're curious how _using_ DeltaEncoding affects things, [this comment](https://github.com/spiraldb/vortex/pull/916#pullrequestreview-2326607030) benchmarks the use of DeltaEncoding for the offsets of the varbin array inside an FSST array.

robert3005 commented on encodings/fastlanes/src/delta/compute.rs:
```diff
@@ -1,5 +1,422 @@
+
+impl ScalarAtFn for DeltaArray {
+    fn scalar_at(&self, index: usize) -> VortexResult<Scalar> {
+        let decompressed = SliceFn::slice(self, index, index + 1)?.into_primitive()?;
```
just `slice(self, index, index+1)` is there a reason you want to do `self.slice(index, index+1)`?

danking replied:
I can't do `self.slice` because both `ArrayCompute` and `SliceFn` are in scope, but I could do `slice(self, ...)`, the function. I'll do that.

I guess it does seem a bit silly (in terms of generated code) to do the whole rigamarole of getting an option, asserting its not `None`, just to call a function that I can statically resolve to `SliceFn::slice`. Thoughts on that?

robert3005 replied:
hm fair, I haven't thought about arraycompute and individual functions colliding in scope. Fair to keep it as is

robert3005 commented on encodings/fastlanes/src/delta/compute.rs:
```diff
@@ -1,5 +1,422 @@
+            physical_start % 1024,
+            limit,
+        )?;
+        if arr.len() != stop - start {
```
This is a weird assert in this place...

danking replied:
As I was developing it was helpful to ensure applying the offset & limit produced the correct length but I'll remove it now.

danking committed:
fix docs

robert3005 commented on encodings/fastlanes/src/delta/compute.rs:
```diff
@@ -1,5 +1,422 @@
+            new_deltas,
+            new_validity,
+            physical_start % 1024,
+            limit,
```
This should just be a number, and 0 should be perfectly fine value for it

danking replied:
Okay, done! The code does seem simpler now.

robert3005 replied:
Can we explicitly specify length here instead of last chunk length?

robert3005 commented on encodings/fastlanes/src/delta/compute.rs:
```diff
@@ -1,5 +1,422 @@
+        })?;
+
+        let new_deltas = deltas.with_dyn(|a| {
+            a.slice()
```
you want `slice(deltas ...)`

robert3005 commented on encodings/fastlanes/src/delta/compute.rs:
```diff
@@ -1,5 +1,422 @@
+        let validity = self.validity();
+        let lanes = self.lanes();
+
+        let new_bases = bases.with_dyn(|a| {
```
you want `slice(bases, ...)`

robert3005 commented on encodings/fastlanes/src/delta/compress.rs:
```diff
@@ -99,13 +112,25 @@ where
     });
-    Ok(decoded)
+    decoded
+        .slice(array.offset(), decoded.len() - array.trailing_garbage())?
```
You can use the original array length here that will be only the valid values, i.e. offset + len

danking replied:
Done.

robert3005 reviewed (changes requested):
There's couple places where we are not using compute functions the right way

danking committed:
remove logical_len and trailing_garbage from metadata

1. logical_len is already stored as the array length.
2. trailing_garbage duplicates the information stored in the logical length.

danking committed:
use the slice function rather than the trait method

danking committed:
eliminate the option for limit

fixup limit

danking committed:
remove length assertion

danking committed:
remove the support for unsigned types in Delta

danking committed:
revert to origin/develop

danking committed:
revert to origin/develop

danking committed:
revert to origin/develop

danking reviewed (commented):
All addressed.

danking committed:
restore origin/develop

danking committed:
directly use logical_len rather than limit

danking committed:
fix up the comment

robert3005 commented on encodings/fastlanes/src/delta/mod.rs:
```diff
@@ -68,7 +162,7 @@ impl DeltaArray {
     pub fn deltas(&self) -> Array {
         self.as_ref()
-            .child(1, self.dtype(), self.len())
+            .child(1, self.dtype(), self.metadata().deltas_len)
```
you could use `deltas_len()` function here

danking replied:
done.

robert3005 commented on encodings/fastlanes/src/delta/mod.rs:
```diff
@@ -19,11 +22,86 @@ impl_encoding!("fastlanes.delta", ids::FL_DELTA, Delta);
 pub struct DeltaMetadata {
     validity: ValidityMetadata,
-    len: usize,
+    deltas_len: usize,
```
As a followup we could get rid of this one, it always have to be ~len rounded up to 1024~ len + offset rounded up to 1024

danking replied:
This isn't true; in fact, this is what took me so long to wrap my head around this PR. Before this PR, if the logical length is not 0 mod 1024, then we encode the last chunk as one vector using scalar operations. After this PR, the above is still  true except it depends on `deltas.len()` (which previously was always equal to the logical length).

robert3005 replied:
oh, I somehow assumed it was the same as fastlanes bitpacking

danking committed:
prefer deltas_len() to metadata().deltas_len
